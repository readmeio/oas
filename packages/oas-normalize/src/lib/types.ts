export interface Options {
  colorizeErrors?: boolean;
  enablePaths?: boolean;
}
