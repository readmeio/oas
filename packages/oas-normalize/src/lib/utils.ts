import YAML, { JSON_SCHEMA } from 'js-yaml';

/**
 * Determine if a given variable is a `Buffer`.
 *
 */
export function isBuffer(obj: any) {
  return (
    obj != null &&
    obj.constructor != null &&
    typeof obj.constructor.isBuffer === 'function' &&
    !!obj.constructor.isBuffer(obj)
  );
}

/**
 * Converts GitHub blob URLs to raw URLs
 */
export function normalizeURL(url: string) {
  if (url.startsWith('https://github.com/') && url.includes('/blob/')) {
    return url.replace('github.com', 'raw.githubusercontent.com').replace('/blob/', '/');
  }
  return url;
}

/**
 * Determine the type of a given variable. Returns `false` if unrecognized.
 *
 */
export function getType(obj: any) {
  if (isBuffer(obj)) {
    return 'buffer';
  } else if (typeof obj === 'object') {
    return 'json';
  } else if (typeof obj === 'string') {
    if (obj.match(/\s*{/)) {
      return 'string-json';
    } else if (obj.match(/\n/)) {
      // Not sure about this...
      return 'string-yaml';
    } else if (obj.substring(0, 4) === 'http') {
      return 'url';
    }

    return 'path';
  }

  return false;
}

/**
 * Determine if a given schema if an OpenAPI definition.
 *
 */
export function isOpenAPI(schema: Record<string, unknown>) {
  return !!schema.openapi;
}

/**
 * Determine if a given schema is a Postman collection.
 *
 * Unfortunately the Postman schema spec doesn't have anything like `openapi` or `swagger` for us
 * to look at but it does require that `info` and `item` be present and as `item` doesn't exist in
 * OpenAPI or Swagger we can use the combination of those two properties to determine if what we
 * have is a Postman collection.
 *
 * @see {@link https://schema.postman.com/json/collection/v2.0.0/collection.json}
 * @see {@link https://schema.postman.com/json/collection/v2.1.0/collection.json}
 */
export function isPostman(schema: Record<string, unknown>): boolean {
  return !!schema.info && !!schema.item;
}

/**
 * Determine if a given schema if an Swagger definition.
 *
 */
export function isSwagger(schema: Record<string, unknown>) {
  return !!schema.swagger;
}

/**
 * Convert a YAML blob or stringified JSON object into a JSON object.
 *
 */
export function stringToJSON(string: Record<string, unknown> | string): Record<string, unknown> {
  if (typeof string === 'object') {
    return string;
  } else if (string.match(/^\s*{/)) {
    return JSON.parse(string);
  }

  return YAML.load(string, { schema: JSON_SCHEMA }) as Record<string, unknown>;
}

/**
 * Determine if a given schema is an API definition that we can support.
 *
 */
export function isAPIDefinition(schema: Record<string, unknown>) {
  return isOpenAPI(schema) || isPostman(schema) || isSwagger(schema);
}

/**
 * Retrieve the type of API definition that a given schema is.
 *
 */
export function getAPIDefinitionType(schema: Record<string, unknown>) {
  if (isOpenAPI(schema)) {
    return 'openapi';
  } else if (isPostman(schema)) {
    return 'postman';
  } else if (isSwagger(schema)) {
    return 'swagger';
  }

  return 'unknown';
}
