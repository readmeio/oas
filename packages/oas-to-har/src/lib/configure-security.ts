import type { AuthForHAR } from './types.js';
import type { OASDocument, SecuritySchemeObject } from 'oas/rmoas.types';

import { isRef } from 'oas/rmoas.types';

function harValue(type: 'cookies' | 'headers' | 'queryString', value: { name: string; value: string }) {
  if (!value.value) return undefined;
  return { type, value };
}

export default function configureSecurity(apiDefinition: OASDocument, values: AuthForHAR, scheme: string) {
  if (!scheme) return undefined;

  if (Object.keys(values || {}).length === 0) return undefined;

  if (!apiDefinition.components.securitySchemes[scheme]) return undefined;
  const security = apiDefinition.components.securitySchemes[scheme] as SecuritySchemeObject & {
    'x-bearer-format'?: string;
  };

  if (isRef(security)) {
    return undefined;
  } else if (!values[scheme]) {
    // If we don't have any data for this auth scheme then we shouldn't add it.
    return false;
  }

  if (security.type === 'http') {
    if (security.scheme === 'basic') {
      const auth = values[scheme];
      if (typeof auth !== 'object') return false;
      if (!auth.user && !auth.pass) return false;

      let user = auth.user ?? null;
      if (user === null || user.length === 0) {
        user = '';
      }

      let pass = auth.pass ?? null;
      if (pass === null || pass.length === 0) {
        pass = '';
      }

      return harValue('headers', {
        name: 'authorization',
        value: `Basic ${Buffer.from(`${user}:${pass}`).toString('base64')}`,
      });
    } else if (security.scheme === 'bearer') {
      return harValue('headers', {
        name: 'authorization',
        value: `Bearer ${values[scheme]}`,
      });
    }
  }

  if (security.type === 'apiKey') {
    if (security.in === 'query') {
      return harValue('queryString', {
        name: security.name,
        value: String(values[scheme]),
      });
    } else if (security.in === 'header') {
      const header = {
        name: security.name,
        value: String(values[scheme]),
      };

      if (security['x-bearer-format']) {
        // Uppercase: token -> Token
        const bearerFormat = security['x-bearer-format'].charAt(0).toUpperCase() + security['x-bearer-format'].slice(1);
        header.name = security.name;
        header.value = `${bearerFormat} ${header.value}`;
      }

      return harValue('headers', header);
    } else if (security.in === 'cookie') {
      return harValue('cookies', {
        name: security.name,
        value: String(values[scheme]),
      });
    }
  }

  if (security.type === 'oauth2') {
    return harValue('headers', {
      name: 'authorization',
      value: `Bearer ${values[scheme]}`,
    });
  }

  return undefined;
}
