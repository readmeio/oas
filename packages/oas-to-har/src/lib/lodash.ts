type Many<T> = T | readonly T[];
type PropertyName = number | string | symbol;
type PropertyPath = Many<PropertyName>;

/**
 * A janky, poorly typed replacement for `lodash.get`.
 *
 * @see {@link https://youmightnotneed.com/lodash#get}
 */
export function get(object: unknown, path?: string): any {
  // If path is not defined or it has false value
  if (!path) return undefined;
  // Check if path is string or array. Regex : ensure that we do not have '.' and brackets.
  // Regex explained: https://regexr.com/58j0k
  const pathArray = String(path).match(/([^[.\]])+/g);
  // Find value
  // @ts-expect-error idk man
  const result = pathArray?.reduce((prevObj, key) => prevObj && prevObj[key], object);
  // If found value is undefined return default value; otherwise return the value
  return result;
}

/**
 * A janky, poorly typed replacement for `lodash.set`.
 *
 * @see {@link https://youmightnotneed.com/lodash#set}
 */
export function set<TResult>(object: object, path: PropertyPath, value: any): TResult {
  // Regex explained: https://regexr.com/58j0k
  const pathArray: PropertyPath | RegExpMatchArray | null = Array.isArray(path)
    ? path
    : String(path).match(/([^[.\]])+/g);

  // @ts-expect-error idk man
  return pathArray?.reduce((acc, key, i) => {
    // @ts-expect-error idk man
    if (acc[key] === undefined) {
      // @ts-expect-error idk man
      acc[key] = {};
    }
    if (i === pathArray.length - 1) {
      // @ts-expect-error idk man
      acc[key] = value;
    }
    // @ts-expect-error idk man
    return acc[key];
  }, object);
}
