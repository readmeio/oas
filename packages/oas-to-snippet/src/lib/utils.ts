import type { Language } from '../types.js';
import type { TargetId } from '@readme/httpsnippet/targets';

import supportedLanguages from '../supportedLanguages.js';

export function getLanguageConfig(lang: Language) {
  let config;
  let language: TargetId;
  let target;

  // If `lang` is an array, then it's a mixture of language and targets like `[php, guzzle]` or
  // `[javascript, axios]` so we need to a bit of work to pull out the necessary information needed
  // to build the snippet. For backwards compatibility sake we still need to support supplying
  // `node-simple` as the language even though `node-simple` does not exist within the list of
  // supported languages.
  if (lang === 'node-simple') {
    config = supportedLanguages.node;
    language = 'node';
    target = 'api';
  } else if (lang === 'curl') {
    config = supportedLanguages.shell;
    language = 'shell';
    target = 'curl';
  } else if (Array.isArray(lang)) {
    if (lang[0] in supportedLanguages) {
      if (lang[1] in supportedLanguages[lang[0]].httpsnippet.targets) {
        config = supportedLanguages[lang[0]];
        language = config.httpsnippet.lang;
        target = lang[1];
      }
    }
  } else if (lang in supportedLanguages) {
    config = supportedLanguages[lang];
    language = config.httpsnippet.lang;
    target = config.httpsnippet.default;
  } else {
    throw new Error('An unknown language was supplied.');
  }

  return {
    config,
    language,
    target,
  };
}

export function getInstallInstructionsForLanguage(
  lang: Language,
  /**
   * For `api` snippet generation we need supply a ReadMe API Registry identifier that'll be used
   * to retrieve the OpenAPI definition to generate their SDK through the `npx api install` command.
   * These users won't be installing the `api` package itself to run the snippets we generate for
   * them.
   *
   * @example @developers/v2.0#17273l2glm9fq4l5
   */
  registryIdentifier?: string,
) {
  const { config, target } = getLanguageConfig(lang);

  const install = config.httpsnippet.targets[target]?.install;
  if (!install) {
    return undefined;
  }

  return install.replace('{registryIdentifier}', registryIdentifier);
}
