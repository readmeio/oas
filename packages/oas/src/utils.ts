import findSchemaDefinition from './lib/find-schema-definition.js';
import matchesMimeType from './lib/matches-mimetype.js';
import { types as jsonSchemaTypes } from './operation/lib/get-parameters-as-json-schema.js';

const supportedMethods = new Set(['get', 'put', 'post', 'delete', 'options', 'head', 'patch', 'trace']);

export { findSchemaDefinition, jsonSchemaTypes, matchesMimeType, supportedMethods };
