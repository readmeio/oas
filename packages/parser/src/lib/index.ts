import type { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

/**
 * Regular expression that matches path parameter templating.
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#path-templating}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#path-templating}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#path-templating}
 */
export const pathParameterTemplateRegExp: RegExp = /\{([^/}]+)}/g;

/**
 * List of HTTP verbs used for OperationItem as per the OpenAPI and Swagger specifications
 *
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#path-item-object}
 * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#path-item-object}
 */
export const supportedHTTPMethods = ['get', 'post', 'put', 'delete', 'patch', 'options', 'head', 'trace'] as const;
export const swaggerHTTPMethods = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch'] as const;

/**
 * Is a given object a Swagger API definition?
 *
 */
export function isSwagger(schema: any): schema is OpenAPIV2.Document {
  return 'swagger' in schema && schema.swagger !== undefined;
}

/**
 * Is a given object an OpenAPI API definition?
 *
 */
export function isOpenAPI(schema: any): schema is OpenAPIV3_1.Document | OpenAPIV3.Document {
  return 'openapi' in schema && schema.openapi !== undefined;
}

/**
 * Is a given object an OpenAPI 3.0 API definition?
 *
 */
export function isOpenAPI30(schema: any): schema is OpenAPIV3.Document {
  return 'openapi' in schema && schema.openapi !== undefined && schema.openapi.startsWith('3.0');
}

/**
 * Is a given object an OpenAPI 3.1 API definition?
 *
 */
export function isOpenAPI31(schema: any): schema is OpenAPIV3_1.Document | OpenAPIV3_1.Document {
  return 'openapi' in schema && schema.openapi !== undefined && schema.openapi.startsWith('3.1');
}

/**
 * Determine the proper name for the API specification schema used by a given schema.
 *
 */
export function getSpecificationName(
  api: OpenAPIV2.Document | OpenAPIV3_1.Document | OpenAPIV3.Document,
): 'OpenAPI' | 'Swagger' {
  return isSwagger(api) ? 'Swagger' : 'OpenAPI';
}
