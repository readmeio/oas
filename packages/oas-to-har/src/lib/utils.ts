import type { JSONSchema, SchemaObject } from 'oas/rmoas.types';

import { get as lodashGet } from 'lodash'; // eslint-disable-line no-restricted-imports

/**
 * Determine if a schema `type` is, or contains, a specific discriminator.
 *
 */
export function hasSchemaType(
  schema: SchemaObject,
  discriminator: 'array' | 'object' | 'string' | 'number' | 'boolean' | 'integer' | 'null',
) {
  if (Array.isArray(schema.type)) {
    return schema.type.includes(discriminator);
  }

  return schema.type === discriminator;
}

/**
 * Because some request body schema shapes might not always be a top-level `properties`, instead
 * nesting it in an `oneOf` or `anyOf` we need to extract the first usable schema that we have. If
 * we don't do this then these non-conventional request body schema payloads may not be properly
 * represented in the HAR that we generate.
 *
 */
export function getSafeRequestBody(obj: any) {
  if ('oneOf' in obj) {
    return getSafeRequestBody(obj.oneOf[0]);
  } else if ('anyOf' in obj) {
    return getSafeRequestBody(obj.anyOf[0]);
  }

  return obj;
}

interface Options {
  parentIsArray?: boolean;
  parentKey?: string;
  payload: unknown;
}

function getSubschemas(schema: any, opts: Options) {
  let subSchemaDataSize = 0;
  if (opts.parentIsArray) {
    // If we don't have data for this parent schema in our body payload then we
    // shouldn't bother spidering further into the schema looking for more `format`s
    // for data that definitely doesn't exist.
    const parentData = lodashGet(opts.payload, opts.parentKey);
    if (parentData === undefined || !Array.isArray(parentData)) {
      return false;
    }

    subSchemaDataSize = parentData.length;
  }

  let subschemas: any[] = [];
  if (subSchemaDataSize > 0) {
    for (let idx = 0; idx < subSchemaDataSize; idx += 1) {
      subschemas = subschemas.concat(
        Object.entries(schema).map(([key, subschema]: [string, JSONSchema]) => ({
          key: opts.parentKey ? [opts.parentKey, idx, key].join('.') : key,
          schema: getSafeRequestBody(subschema),
        })),
      );
    }
  } else {
    subschemas = Object.entries(schema).map(([key, subschema]: [string, JSONSchema]) => ({
      key: opts.parentKey ? [opts.parentKey, key].join('.') : key,
      schema: getSafeRequestBody(subschema),
    }));
  }

  return subschemas;
}

/**
 * With a supplied JSON Schema object, spider through it for any schemas that may contain specific
 * kind of `format` that also happen to be within the current `requestBody` payload that we're
 * creating a HAR representation for.
 *
 */
export function getTypedFormatsInSchema(
  format: 'json' | 'binary',
  schema: any,
  opts: Options,
): boolean | string | (string | boolean)[] {
  try {
    if (schema?.format === format) {
      if (opts.parentIsArray) {
        const parentData = lodashGet(opts.payload, opts.parentKey);
        if (parentData !== undefined && Array.isArray(parentData)) {
          return Object.keys(parentData)
            .map(pdk => {
              const currentKey = [opts.parentKey, pdk].join('.');
              if (lodashGet(opts.payload, currentKey) !== undefined) {
                return currentKey;
              }

              return false;
            })
            .filter(Boolean);
        }
      } else if (opts.parentKey && lodashGet(opts.payload, opts.parentKey) !== undefined) {
        return opts.parentKey;
      } else if (opts.payload !== undefined) {
        // If this payload is present and we're looking for a specific format then we should assume
        // that the **root** schema of the request body is that format, and we aren't trafficking in
        // a nested object or array schema.
        return true;
      }

      return false;
    }

    const subschemas = getSubschemas(schema, opts);
    if (!subschemas) {
      return false;
    }

    return subschemas
      .map(({ key, schema: subschema }) => {
        if ('properties' in subschema) {
          return getTypedFormatsInSchema(format, subschema.properties, { payload: opts.payload, parentKey: key });
        } else if ('items' in subschema) {
          if ((subschema.items as JSONSchema)?.properties) {
            return getTypedFormatsInSchema(format, (subschema.items as JSONSchema).properties, {
              payload: opts.payload,
              parentKey: key,
              parentIsArray: true,
            });
          }

          return getTypedFormatsInSchema(format, subschema.items, {
            payload: opts.payload,
            parentKey: key,
            parentIsArray: true,
          });
        }

        // If this schema has neither `properties` or `items` then it's a regular schema
        // we can re-run.
        return getTypedFormatsInSchema(format, subschema, { payload: opts.payload, parentKey: key });
      })
      .flat()
      .filter(Boolean);
  } catch (err) {
    // If this fails for whatever reason then we should act as if we didn't find any `format`'d
    // schemas.
    return [];
  }
}
