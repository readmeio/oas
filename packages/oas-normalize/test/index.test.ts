import type { OpenAPIV3 } from 'openapi-types';

import fs from 'node:fs';
import path from 'node:path';

import fetchMock from 'fetch-mock';
import { describe, afterEach, beforeAll, beforeEach, it, expect } from 'vitest';

import OASNormalize, { getAPIDefinitionType, isAPIDefinition } from '../src';
import { isOpenAPI, isPostman, isSwagger } from '../src/lib/utils';

function cloneObject(obj) {
  return JSON.parse(JSON.stringify(obj));
}

describe('#load', () => {
  describe.each([
    ['Swagger 2.0', '2.0'],
    ['OpenAPI 3.0', '3.0'],
    ['OpenAPI 3.1', '3.1'],
  ])('%s support', (_, version) => {
    let json;
    let yaml;

    beforeEach(async () => {
      json = await import(`@readme/oas-examples/${version}/json/petstore.json`).then(r => r.default);
      yaml = require.resolve(`@readme/oas-examples/${version}/yaml/petstore.yaml`);
    });

    it('should reject if unrecognized file supplied', async () => {
      await expect(new OASNormalize(cloneObject).load()).rejects.toThrow('Could not load this file.');
    });

    it('should support JSON objects', async () => {
      const o = new OASNormalize(cloneObject(json));
      await expect(o.load()).resolves.toStrictEqual(json);
    });

    it('should support stringified JSON objects', async () => {
      const def = JSON.stringify(cloneObject(json));
      const o = new OASNormalize(def);

      await expect(o.load()).resolves.toStrictEqual(json);
    });

    it('should support YAML', async () => {
      const o = new OASNormalize(fs.readFileSync(yaml, 'utf8'));
      await expect(o.load()).resolves.toStrictEqual(json);
    });

    it('should support buffers', async () => {
      const o = new OASNormalize(fs.readFileSync(yaml));
      await expect(o.load()).resolves.toStrictEqual(json);
    });

    describe('urls', () => {
      afterEach(() => {
        fetchMock.restore();
      });

      it('should support URLs', async () => {
        fetchMock.get(`http://example.com/api-${version}.json`, json);
        const o = new OASNormalize(`http://example.com/api-${version}.json`);

        await expect(o.load()).resolves.toStrictEqual(json);
      });

      it('should support HTTPS URLs', async () => {
        fetchMock.get(`https://example.com/api-${version}.json`, json);
        const o = new OASNormalize(`https://example.com/api-${version}.json`);

        await expect(o.load()).resolves.toStrictEqual(json);
      });

      it('should convert GitHub repo URLs to raw URLs', async () => {
        fetchMock.get('https://raw.githubusercontent.com/readmeio/oas-examples/main/3.0/json/petstore.json', json);
        const o = new OASNormalize('https://github.com/readmeio/oas-examples/blob/main/3.0/json/petstore.json');

        await expect(o.load()).resolves.toStrictEqual(json);
      });
    });

    describe('paths', () => {
      it('should support paths', async () => {
        const o = new OASNormalize(yaml, { enablePaths: true });
        await expect(o.load()).resolves.toStrictEqual(json);
      });

      it('should reject if `enablePaths` is not set', async () => {
        const o = new OASNormalize(yaml);
        await expect(o.load()).rejects.toThrow('Use `opts.enablePaths` to enable accessing local files.');
      });
    });
  });

  describe('Postman support', () => {
    it('should be able to load a Postman collection', async () => {
      const postman = await import('./__fixtures__/postman/petstore.collection.json').then(r => r.default);
      const o = new OASNormalize(postman);
      await expect(o.load()).resolves.toStrictEqual(
        expect.objectContaining({
          info: expect.objectContaining({
            // eslint-disable-next-line camelcase
            _postman_id: '0b2e8577-2899-4229-bb1c-4cb031108c2f',
          }),
        }),
      );
    });
  });

  describe('quirks', () => {
    it('should not convert date strings in YAML files into Date objects', async () => {
      const yaml = require.resolve('./__fixtures__/quirks/yaml-date.yaml');
      const o = new OASNormalize(fs.readFileSync(yaml, 'utf8'));

      const s = await o.load();
      expect(typeof s.info.version).toBe('string');
    });
  });
});

describe('#bundle', () => {
  it('should bundle an external schema in', async () => {
    const petSchema = await import('./__fixtures__/bundle/pet.json').then(r => r.default);
    const contents = require.resolve('./__fixtures__/bundle/definition.json');
    const o = new OASNormalize(contents, { enablePaths: true });
    const bundled = (await o.bundle()) as OpenAPIV3.Document;

    expect(bundled.components?.requestBodies?.Pet).toStrictEqual({
      description: 'Pet object that needs to be added to the store',
      required: true,
      content: {
        'application/json': {
          schema: {
            $ref: '#/components/requestBodies/Pet/content/application~1xml/schema',
          },
        },
        'application/xml': {
          schema: petSchema,
        },
      },
    });
  });

  describe('Postman support', () => {
    it('should convert a Postman collection if supplied', async () => {
      const postman = await import('./__fixtures__/postman/petstore.collection.json').then(r => r.default);
      const o = new OASNormalize(postman);
      const bundled = (await o.bundle()) as OpenAPIV3.Document;

      // There's nothing to bundle in a Postman collection so we're really just testing here if it
      // was properly converted to an OpenAPI definition.
      expect(bundled.components).toStrictEqual({
        securitySchemes: {
          apikeyAuth: { type: 'http', scheme: 'apikey' },
          oauth2Auth: { type: 'http', scheme: 'oauth2' },
        },
      });
    });
  });
});

describe('#deref', () => {
  it('should dereference a definition', async () => {
    const openapi = await import('@readme/oas-examples/3.0/json/petstore.json').then(r => r.default);
    expect(openapi.paths['/pet'].post.requestBody).toStrictEqual({
      $ref: '#/components/requestBodies/Pet',
    });

    const o = new OASNormalize(cloneObject(openapi));
    const deref = (await o.deref()) as OpenAPIV3.Document;

    expect(deref?.paths?.['/pet']?.post?.requestBody).toStrictEqual({
      description: 'Pet object that needs to be added to the store',
      required: true,
      content: {
        'application/json': expect.any(Object),
        'application/xml': expect.any(Object),
      },
    });
  });

  describe('Postman support', () => {
    it('should convert a Postman collection if supplied', async () => {
      const postman = await import('./__fixtures__/postman/petstore.collection.json').then(r => r.default);

      const o = new OASNormalize(postman);
      const deref = (await o.deref()) as OpenAPIV3.Document;

      expect(deref?.paths?.['/v2/pet']?.post?.requestBody).toStrictEqual({
        content: {
          'application/json': {
            schema: {
              type: 'object',
              example: expect.objectContaining({
                name: 'doggie',
              }),
            },
          },
        },
      });
    });
  });
});

describe('#validate', () => {
  it("should not convert a Swagger definition to OpenAPI if we don't want to", async () => {
    const swagger = await import('@readme/oas-examples/2.0/json/petstore.json').then(r => r.default);
    const o = new OASNormalize(cloneObject(swagger));

    await expect(o.validate()).resolves.toStrictEqual(swagger);
  });

  it('should error out on a definition a missing component', async () => {
    const contents = path.join(__dirname, '__fixtures__', 'invalid', 'swagger.json');
    const o = new OASNormalize(contents, { enablePaths: true });

    await expect(o.validate()).rejects.toThrow('Token "Category" does not exist.');
  });

  it('should error if a schema is missing', async () => {
    const contents = path.join(__dirname, '__fixtures__', 'invalid', 'openapi.json');
    const o = new OASNormalize(contents, { enablePaths: true });

    await expect(o.validate()).rejects.toThrow('Token "Error" does not exist.');
  });

  it("should error out when a definition doesn't match the spec", async () => {
    // This definition is missing `paths`, which should incur a failed validation check.
    const contents = {
      openapi: '3.0.3',
      info: {
        title: 'Example OpenAPI base file for `oas`.',
        version: '1.0',
      },
    };

    const o = new OASNormalize(contents);
    await expect(o.validate()).rejects.toThrow('Supplied schema is not a valid OpenAPI definition.');
  });

  it("should error out when a definition doesn't match the schema", async () => {
    const o = new OASNormalize(require.resolve('./__fixtures__/invalid/openapi-3.1.json'), { enablePaths: true });

    await expect(o.validate()).rejects.toStrictEqual(
      expect.objectContaining({
        message: expect.stringContaining("REQUIRED must have required property 'name'"),
        details: expect.any(Array),
      }),
    );
  });

  it('should error out, and show all errors, when a definition has lots of problems', async () => {
    const o = new OASNormalize(require.resolve('./__fixtures__/invalid/openapi-very-invalid.json'), {
      enablePaths: true,
    });

    await expect(o.validate()).rejects.toMatchSnapshot();
  });

  it('should error out for empty file', async () => {
    const o = new OASNormalize(require.resolve('./__fixtures__/invalid/empty.json'), {
      enablePaths: true,
    });

    await expect(o.validate()).rejects.toStrictEqual(new Error('No file contents found.'));
  });

  // Skipping because the `chalk` dependency of `better-ajv-errors` within `openapi-parser` has
  // issues in CI. Test works fine locally though!
  // eslint-disable-next-line vitest/no-disabled-tests
  it.skip('should colorize errors when `opts.colorizeErrors` is present', async () => {
    const o = new OASNormalize(require.resolve('./__fixtures__/invalid/openapi-3.1.json'), {
      colorizeErrors: true,
      enablePaths: true,
    });

    await expect(o.validate()).rejects.toMatchSnapshot();
  });

  describe.each([
    ['Swagger 2.0', '2.0'],
    ['OpenAPI 3.0', '3.0'],
    ['OpenAPI 3.1', '3.1'],
  ])('%s support', (_, version) => {
    afterEach(() => {
      fetchMock.restore();
    });

    it('should validate a URL hosting JSON as expected', async () => {
      const json = await import(`@readme/oas-examples/${version}/json/petstore.json`).then(r => r.default);

      fetchMock.get(`http://example.com/api-${version}.json`, cloneObject(json));
      const o = new OASNormalize(`http://example.com/api-${version}.json`);

      await expect(o.validate({ convertToLatest: true })).resolves.toMatchSnapshot();
    });

    it('should validate a JSON path as expected', async () => {
      const o = new OASNormalize(require.resolve(`@readme/oas-examples/${version}/json/petstore.json`), {
        enablePaths: true,
      });

      await expect(o.validate({ convertToLatest: true })).resolves.toMatchSnapshot();
    });

    it('should validate a URL hosting YAML as expected', async () => {
      const yaml = fs.readFileSync(require.resolve(`@readme/oas-examples/${version}/yaml/petstore.yaml`), 'utf8');
      fetchMock.get(`http://example.com/api-${version}.yaml`, yaml);
      const o = new OASNormalize(`http://example.com/api-${version}.yaml`);

      await expect(o.validate({ convertToLatest: true })).resolves.toMatchSnapshot();
    });

    it('should validate a YAML path as expected', async () => {
      const o = new OASNormalize(require.resolve(`@readme/oas-examples/${version}/yaml/petstore.yaml`), {
        enablePaths: true,
      });

      await expect(o.validate({ convertToLatest: true })).resolves.toMatchSnapshot();
    });
  });

  describe('Postman support', () => {
    it('should support converting a Postman collection to OpenAPI (validating it in the process)', async () => {
      const o = new OASNormalize(require.resolve('./__fixtures__/postman/petstore.collection.json'), {
        enablePaths: true,
      });

      await expect(o.validate({ convertToLatest: true })).resolves.toMatchSnapshot();
    });
  });
});

describe('#version', () => {
  it('should detect an OpenAPI definition', async () => {
    await expect(
      new OASNormalize(require.resolve('@readme/oas-examples/3.0/json/petstore.json'), { enablePaths: true }).version(),
    ).resolves.toStrictEqual({
      specification: 'openapi',
      version: '3.0.0',
    });

    await expect(
      new OASNormalize(require.resolve('@readme/oas-examples/3.1/json/petstore.json'), { enablePaths: true }).version(),
    ).resolves.toStrictEqual({
      specification: 'openapi',
      version: '3.1.0',
    });
  });

  it('should detect a Postman collection', async () => {
    await expect(
      new OASNormalize(require.resolve('./__fixtures__/postman/petstore.collection.json'), {
        enablePaths: true,
      }).version(),
    ).resolves.toStrictEqual({
      specification: 'postman',
      version: '2.1.0',
    });
  });

  it('should detect a Swagger definition', async () => {
    await expect(
      new OASNormalize(require.resolve('@readme/oas-examples/2.0/json/petstore.json'), { enablePaths: true }).version(),
    ).resolves.toStrictEqual({
      specification: 'swagger',
      version: '2.0',
    });
  });
});

describe('#utils', () => {
  let openapi;
  let postman;
  let swagger;

  beforeAll(async () => {
    openapi = await import('@readme/oas-examples/3.0/json/petstore.json').then(r => r.default);
    postman = await import('./__fixtures__/postman/petstore.collection.json').then(r => r.default);
    swagger = await import('@readme/oas-examples/2.0/json/petstore.json').then(r => r.default);
  });

  describe('#isAPIDefinition / #getAPIDefinitionType', () => {
    it('should identify an OpenAPI definition', () => {
      expect(isAPIDefinition(openapi)).toBe(true);
      expect(getAPIDefinitionType(openapi)).toBe('openapi');
    });

    it('should identify a Postman definition', () => {
      expect(isAPIDefinition(postman)).toBe(true);
      expect(getAPIDefinitionType(postman)).toBe('postman');
    });

    it('should identify a Swagger definition', () => {
      expect(isAPIDefinition(swagger)).toBe(true);
      expect(getAPIDefinitionType(swagger)).toBe('swagger');
    });

    it('should not identify a non-API definition as one', async () => {
      const pkg = await import('../package.json').then(r => r.default);

      expect(isAPIDefinition(pkg)).toBe(false);
      expect(getAPIDefinitionType(pkg)).toBe('unknown');
    });
  });

  describe('#isOpenAPI', () => {
    it('should identify an OpenAPI definition', () => {
      expect(isOpenAPI(openapi)).toBe(true);
    });

    it('should not misidentify a Swagger definition', () => {
      expect(isOpenAPI(swagger)).toBe(false);
    });

    it('should not misidentify a Postman collection', () => {
      expect(isOpenAPI(postman)).toBe(false);
    });
  });

  describe('#isPostman', () => {
    it('should identify a Postman collection', () => {
      expect(isPostman(postman)).toBe(true);
    });

    it('should not misidentify a Swagger definition', () => {
      expect(isPostman(swagger)).toBe(false);
    });

    it('should not misidentify an OpenAPI', () => {
      expect(isPostman(openapi)).toBe(false);
    });
  });

  describe('#isSwagger', () => {
    it('should identify a Swagger definition', () => {
      expect(isSwagger(swagger)).toBe(true);
    });

    it('should not misidentify an OpenAPI definition', () => {
      expect(isSwagger(openapi)).toBe(false);
    });

    it('should not misidentify a Postman collection', () => {
      expect(isSwagger(postman)).toBe(false);
    });
  });
});
