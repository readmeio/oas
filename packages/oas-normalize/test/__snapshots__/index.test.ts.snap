// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#validate > should error out, and show all errors, when a definition has lots of problems 1`] = `
"OpenAPI schema validation failed.

REQUIRED must have required property 'url'

   7 |   },
   8 |   "servers": [
>  9 |     {
     |     ^ ☹️  url is missing here!
  10 |       "urll": "http://petstore.swagger.io/v2"
  11 |     }
  12 |   ],

ADDITIONAL PROPERTY must NOT have additional properties

  24 |     "/pet/findByStatus": {
  25 |       "get": {
> 26 |         "tagss": [
     |         ^^^^^^^ 😲  tagss is not expected to be here!
  27 |           "pet"
  28 |         ],
  29 |         "summary": "Finds Pets by status","
`;

exports[`#validate > should error out, and show all errors, when a definition has lots of problems 2`] = `
[
  {
    "instancePath": "/servers/0",
    "keyword": "required",
    "message": "must have required property 'url'",
    "params": {
      "missingProperty": "url",
    },
    "schemaPath": "#/required",
  },
  {
    "instancePath": "/servers/0",
    "keyword": "additionalProperties",
    "message": "must NOT have additional properties",
    "params": {
      "additionalProperty": "urll",
    },
    "schemaPath": "#/additionalProperties",
  },
  {
    "instancePath": "/paths/~1pet~1findByStatus/get",
    "isIdentifierLocation": true,
    "keyword": "additionalProperties",
    "message": "must NOT have additional properties",
    "params": {
      "additionalProperty": "tagss",
    },
    "schemaPath": "#/additionalProperties",
  },
]
`;
